### DISCUSSION TITLE
How to Add additional PHY drivers to the Linux Kernel

### DISCUSSION THREAD

**Discussion by ryanl01** [polarfire-soc/polarfire-soc-documentation#229] [How to] [answered]
Hello,
Does anyone know what changes need to be made to add support for a different PHY device? I have made the hardwire connections to the icicle kit for testing but realized the linux drivers will also need to be updated to support this.

Thanks in advance,
Ryan

**Comment by vfalanis** [comment]
Hi Ryan,

Asumming you mean an ethernet PHY, I would suggest to have a look at the reg property in the phy stanza in the device tree and update the property based on address used by your PHY device.

Normally, it the address is set correctly it should use the generic drivers to identify the device.

Hope this helps

**Reply by ryanl01** [reply]
Hello Valentina,
Correct. Due to part availability issues, I am trying to make a new PHY device function with the ICICLE development kit. We have its address wired to be the same address as the existing PHY. When linux boots, it can attach ETH1 since it matches the PHY address, but it is unable to obtain an IP address. I assumed this was due to the proper drivers not being in the linux kernel that was built for the development kit. Do you have any insight on this/how to navigate this issue? I am new this kind of development and it appears I am jumping into quite the large pool.

Thanks,
Ryan

**Reply by vfalanis** [reply]
Hi Ryan,

Were you able to resolve this issue or are you still looking into it? Normally it depends on the PHY. You could perhaps look at the drivers/net/phy directory in the Linux kernel to inspect the available configs and drivers and see if there is something you need to enable for the particular PHY you are looking for.

Thanks

**Reply by ryanl01** [reply, marked as answer]
Hi Valentina,

Sorry, I forgot to update this thread. I was able to confirm the new PHY working with the ICICLE development kit. The issue was in the PHY address that was set on the new PHY. The ICICLE kit has a dual PHY with addresses 0x08 and 0x09. I had utilized 0x09 on the new device thinking they would behave the same on the ICICLE kit, but unfortunately the support for those addresses is not the same. On some level of inspiration (I can't remember what I found that made me think it would work), I changed the new device's PHY address to be 0x08 instead and the device just worked with the normal kernel. I still have not been able to find out why the other PHY address (0x09) was not functional, and I will need to continue looking into this.

Thanks,
Ryan

**Reply by TobiasPaulikat** [reply]
Hi Everyone, 
Can you explain me where I can find the PHY Adress ? 
I want to implement H.264 to Ethernet from the Video-Kit to the Discovery-Kit, but while booting the System I receive an error, that the Phy could not be attached.

Thanks

**Reply by hughbreslin** [reply]
@TobiasPaulikat there is a table in the [schematics](https://ww1.microchip.com/downloads/aemDocuments/documents/FPGA/ProductDocuments/ReferenceManuals/PolarFire_SoC_Discovery_Kit_Rev2_SCH_090423_01.pdf) on page 6 which shows the PHY configuration:

![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/44862881/f1150160-7c8d-43b0-ad30-5bb3f56392ba)

In this case the PHY address is 1011 in binary or 11 in decimal.

You can also see this in the [device tree](https://github.com/linux4microchip/linux/blob/linux-6.1-mchp/arch/riscv/boot/dts/microchip/mpfs-disco-kit.dts#L152)

**Reply by TobiasPaulikat** [reply]
Thanks Hugh, 
I can't find the .dts for the Discokit before.

**Reply by keck9939** [reply]
While the dts contains the entry to enable Ethernet on the Disco, it does not seem to be enabled in the prebuilt 2024.02 kernel (there is no Ethernet and if I dump the /sys/firmware/devicetree/base using dtc, the ethernet device shows up disabled. Interestingly Ethernet does function in u-boot. Since the kernel itself seems to have ethernet support e.g. dhcp is running, this seems somewhat strange. Is there a simple way to get Ethernet enabled or am I going to have to dig in and figure out how to use Yocto?

**Reply by keck9939** [reply]
After digging in and installing Yocto, dumping lots of dtb files, I finally figured out that although I thought I had the absolute latest version of the kernel, I evidently did not. The 2024.02.1 version obtained from this location: https://github.com/polarfire-soc/meta-polarfire-soc-yocto-bsp/releases/tag/v2024.02.1 does have functioning Ethernet.
